x/bank/keeper.go

```diff
@@ -8,63 +8,59 @@ import (
-	am sdk.AccountMapper
+func getCoins(ctx sdk.Context, am sdk.AccountMapper, addr sdk.Address) sdk.Coins {
+	acc := am.GetAccount(ctx, addr)
+	if acc == nil {
+		return sdk.Coins{}
+	}
+	return acc.GetCoins()
 }
 
-// NewCoinKeeper returns a new CoinKeeper
-func NewCoinKeeper(am sdk.AccountMapper) CoinKeeper {
-	return CoinKeeper{am: am}
+func setCoins(ctx sdk.Context, am sdk.AccountMapper, addr sdk.Address, amt sdk.Coins) sdk.Error {
+	acc := am.GetAccount(ctx, addr)
+	if acc == nil {
+		acc = am.NewAccountWithAddress(ctx, addr)
+	}
+	acc.SetCoins(amt)
+	am.SetAccount(ctx, acc)
+	return nil
 }
 
-// GetCoins returns the coins at the addr.
-func (ck CoinKeeper) GetCoins(ctx sdk.Context, addr sdk.Address, amt sdk.Coins) sdk.Coins {
-	acc := ck.am.GetAccount(ctx, addr)
-	return acc.GetCoins()
+// HasCoins returns whether or not an account has at least amt coins.
+func hasCoins(ctx sdk.Context, am sdk.AccountMapper, addr sdk.Address, amt sdk.Coins) bool {
+	return getCoins(ctx, am, addr).IsGTE(amt)
 }
 
 // SubtractCoins subtracts amt from the coins at the addr.
-func (ck CoinKeeper) SubtractCoins(ctx sdk.Context, addr sdk.Address, amt sdk.Coins) (sdk.Coins, sdk.Error) {
-	acc := ck.am.GetAccount(ctx, addr)
-	if acc == nil {
-		return amt, sdk.ErrUnknownAddress(addr.String())
-	}
-
-	coins := acc.GetCoins()
-	newCoins := coins.Minus(amt)
+func subtractCoins(ctx sdk.Context, am sdk.AccountMapper, addr sdk.Address, amt sdk.Coins) (sdk.Coins, sdk.Error) {
+	oldCoins := getCoins(ctx, am, addr)
+	newCoins := oldCoins.Minus(amt)
 	if !newCoins.IsNotNegative() {
-		return amt, sdk.ErrInsufficientCoins(fmt.Sprintf("%s < %s", coins, amt))
+		return amt, sdk.ErrInsufficientCoins(fmt.Sprintf("%s < %s", oldCoins, amt))
 	}
-
-	acc.SetCoins(newCoins)
-	ck.am.SetAccount(ctx, acc)
-	return newCoins, nil
+	err := setCoins(ctx, am, addr, newCoins)
+	return newCoins, err
 }
 
 // AddCoins adds amt to the coins at the addr.
-func (ck CoinKeeper) AddCoins(ctx sdk.Context, addr sdk.Address, amt sdk.Coins) (sdk.Coins, sdk.Error) {
-	acc := ck.am.GetAccount(ctx, addr)
-	if acc == nil {
-		acc = ck.am.NewAccountWithAddress(ctx, addr)
+func addCoins(ctx sdk.Context, am sdk.AccountMapper, addr sdk.Address, amt sdk.Coins) (sdk.Coins, sdk.Error) {
+	oldCoins := getCoins(ctx, am, addr)
+	newCoins := oldCoins.Plus(amt)
+	if !newCoins.IsNotNegative() {
+		return amt, sdk.ErrInsufficientCoins(fmt.Sprintf("%s < %s", oldCoins, amt))
 	}
-
-	coins := acc.GetCoins()
-	newCoins := coins.Plus(amt)
-
-	acc.SetCoins(newCoins)
-	ck.am.SetAccount(ctx, acc)
-	return newCoins, nil
+	err := setCoins(ctx, am, addr, newCoins)
+	return newCoins, err
 }
 
 // SendCoins moves coins from one account to another
-func (ck CoinKeeper) SendCoins(ctx sdk.Context, fromAddr sdk.Address, toAddr sdk.Address, amt sdk.Coins) sdk.Error {
-	_, err := ck.SubtractCoins(ctx, fromAddr, amt)
+func sendCoins(ctx sdk.Context, am sdk.AccountMapper, fromAddr sdk.Address, toAddr sdk.Address, amt sdk.Coins) sdk.Error {
```

Comment:
Can we be explicit (in the comment) that if an error is returned, the caller *must* revert state changes (otherwise coins could be subtracted but not added).

Reply:
Or at least the CacheWrap should be thrown out.  That should be the general behavior upon error.

Reply:
I thought this was going to be handled by BaseApp or something?

Reply:
It is - just make it clear in the comment so if someone writes a different baseapp they don't accidentally destroy coins